introspect_test.go

```diff
@@ -0,0 +1,190 @@
+package libp2p
+
+import (
+	"context"
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/websocket"
+	"github.com/libp2p/go-libp2p-core/introspect"
+	"github.com/libp2p/go-libp2p-core/metrics"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-introspection/introspection"
+	"github.com/stretchr/testify/require"
+	"net/url"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestIntrospector(t *testing.T) {
+	msg1 := []byte("1")
+	msg2 := []byte("12")
+	msg3 := []byte("111")
+	msg4 := []byte("0000")
+
+	iaddr := "0.0.0.0:9999"
+	ctx := context.Background()
+
+	// create host 1 with introspector
+	h1, err := New(ctx, Introspector(introspection.NewDefaultIntrospector(iaddr)), BandwidthReporter(metrics.NewBandwidthCounter()))
+	require.NoError(t, err)
+	defer h1.Close()
+
+	// create host 2
+	h2, err := New(ctx)
+	defer h2.Close()
+
+	// create host 3
+	h3, err := New(ctx)
+	defer h3.Close()
+
+	// host1 -> CONNECTS -> host2
+	require.NoError(t, h1.Connect(ctx, peer.AddrInfo{h2.ID(), h2.Addrs()}))
+
+	// host3 -> CONNECTS -> host1
+	require.NoError(t, h3.Connect(ctx, peer.AddrInfo{h1.ID(), h1.Addrs()}))
+
+	// host1 -> OPENS STREAM 1 -> host3, Writes a message & then reads the response
+	var wg sync.WaitGroup
+	p1 := protocol.ID("h1h3")
+	h3.SetStreamHandler(p1, func(s network.Stream) {
+		bz := make([]byte, len(msg1))
+		_, err := s.Read(bz)
+		require.NoError(t, err)
+		_, err = s.Write(msg2)
+		require.NoError(t, err)
+		wg.Done()
+	})
+	s1, err := h1.NewStream(ctx, h3.ID(), p1)
+	require.NoError(t, err)
+	wg.Add(1)
+	_, err = s1.Write(msg1)
+	require.NoError(t, err)
+	bz1 := make([]byte, len(msg2))
+	wg.Wait()
+	_, err = s1.Read(bz1)
+	require.NoError(t, err)
```

Comment:
You can do `require := require.New(t)` at the top, and omit the first `t` parameter from all assertions.